core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -1810,16 +1810,17 @@ protected Type visitInPredicate(InPredicate node, StackableAstVisitorContext<Con
             process(value, context);
 
             Expression valueList = node.getValueList();
-            process(valueList, context);
-
             if (valueList instanceof InListExpression) {
+                process(valueList, context);
                 InListExpression inListExpression = (InListExpression) valueList;
 
                 coerceToSingleType(context,
                         "IN value and list items must be the same type: %s",
                         ImmutableList.<Expression>builder().add(value).addAll(inListExpression.getValues()).build());
             }
             else if (valueList instanceof SubqueryExpression) {
+                subqueryInPredicates.add(NodeRef.of(node));
+                setExpressionType(valueList, analyzeSubquery((SubqueryExpression) valueList, context));
                 coerceToSingleType(context, node, "value and result of subquery must be of the same type for IN expression: %s vs %s", value, valueList);
             }
```

Comment:
else throw?

Reply:
Yes, but that's unrelated to this PR. I'll add a separate commit.